### PR TITLE
Adds split functions for sequence types including lists, vectors, seqs and strings

### PR DESCRIPTION
**Pull request by Izaakwltn**
This adds splitting by single-character delimiters

### CODE REVIEW

File: library/string.lisp [hunk 1]
```diff
@@ -134,6 +136,32 @@ does not have that suffix."
   (define (chars str)
     "Returns an iterator over the characters in `str`."
     (iter:into-iter str))
+  
+  (declare delim-split (Char -> String -> (List String)))
```
Comment (stylewarning) [inline library/string.lisp:140]: can we just call it split?

Comment (stylewarning) [inline library/string.lisp:140]: it might be desirable to return an iterator

**Review by stylewarning — COMMENTED** [review]
in the future it might be nice for this to work on any sequence type with any Eq element type

File: library/string.lisp (hunk 1 above)
Comment (Izaakwltn) [reply library/string.lisp:140]: I was thinking about that- I'm not sure whether the other split is useful or even desirable.

**Comment by Izaakwltn** [comment]
A few updates on the most recent commit:

- Added the `splittable` class which has a method `split` [`coalton-library/split`]
- Added `splittable` instances for `List` `Vector` and `Seq` containing elements with `Eq` instances.
- Renamed the existing `string:split` to `string:bisect`
- The new `string:split` calls `split:split` on a list of chars.

**Comment by stylewarning** [comment]
I'm not 100% convinced this is a class we should include. My main reason is that it seems a little contrived: we wanted something generic so we forced it to be. I think the true generic aspect of something that splittable isn't that it's splittable, but rather that elements have a natural ordering (i.e., they can be indexed) and each element can be accessed in finite time.

My recommendation for the direction of this PR is to create a function `split` in each relevant module (no need to give distinguishing names), forgo the type class, and we can ponder in the future some nice classes to capture such sequential properties.

**Comment by Izaakwltn** [comment]
Now this merely defines a function split for string, seq, list, and vector.

**Comment by Izaakwltn** [comment]
No urgency, but ready for review whenever the next PR/Issue cleanup happens.

5/10:
- rebased onto main so it's still compatible
- cleaned up formatting

File: library/seq.lisp [hunk 2]
```diff
@@ -207,6 +209,30 @@ a new `Seq` instance."
                    (rebalance-branches
                     (fold <> (butlast lsubts) (make-list nsubts (butfirst rsubts))))))))))))
 
+  (declare split ((Eq :a) (types:RuntimeRepr :a) => :a -> (Seq :a) -> (iter:Iterator (Seq :a))))
```
Comment (eliaslfox) [inline library/seq.lisp:212]: This implementation needs to be streaming. The current version won't work with infinite iterators.

File: library/vector.lisp [hunk 3]
```diff
@@ -247,6 +248,18 @@
...
 
+  (declare list->vec (List :a -> Vector :a))
+  (define (list->vec x)
+    (into x))
+
+  (declare vec->list (Vector :a -> List :a))
+  (define (vec->list x)
+    (into x))
+
+  (declare split ((Eq :a) => :a -> (Vector :a) -> (iter:Iterator (Vector :a))))
```
Comment (eliaslfox) [inline library/vector.lisp:259]: Write an implementation that doesn't require converting to lists and back.

File: library/vector.lisp [hunk 4]
```diff
@@ -247,6 +248,18 @@
      iter)
     Unit)
 
+  (declare list->vec (List :a -> Vector :a))
```
Comment (eliaslfox) [inline library/vector.lisp:251]: don't define this

File: library/vector.lisp [hunk 5]
```diff
@@ -247,6 +248,18 @@
...
+  (define (list->vec x)
+    (into x))
+
+  (declare vec->list (Vector :a -> List :a))
```
Comment (eliaslfox) [inline library/vector.lisp:255]: don't define this

File: library/seq.lisp (hunk 2 above)
Comment (Izaakwltn) [reply library/seq.lisp:212]: I'm confused about this- 

I don't believe`Seq` can be infinite, so I'm not sure how a `seq` could or should generate an infinite/stream `Iterator`.

Comment (Izaakwltn) [reply library/seq.lisp:212]: Or do you mean simply that I should also define split for `Iterator` and ensure infinite compatibility?

**Comment by eliaslfox** [comment]
Add iter split:

```
(iter:for-each log
  (iter:split 3
  (map
    (fn (x)
      (log x)
      x)
    (iter:into-iter (1 2 3 2 3 1)))))


1
2
3
(1 2)
2
3
(2)
1
(1)
```

File: library/list.lisp [hunk 6]
```diff
@@ -764,5 +780,7 @@ This function is equivalent to all size-N elements of `(COMBS L)`."
   (define-instance (Default (List :a))
     (define (default) Nil)))
 
+
```
Comment (eliaslfox) [inline library/list.lisp:783]: ?

**Comment by eliaslfox** [comment]
Read [this](https://wiki.haskell.org/Bracket_pattern) and implement it for Coalton w/ `unwind-protect`.

**Review by stylewarning — CHANGES_REQUESTED** [review]
some humble requests

**Comment by Izaakwltn** [comment]
I'm going to close this for now

**Pull request closed without merging**